src/treq/api.py

```diff
@@ -116,19 +120,19 @@ def request(method, url, **kwargs):
     :rtype: Deferred that fires with an IResponse provider.
 
     """
-    return _client(**kwargs).request(method, url, **kwargs)
+    return _client(kwargs).request(method, url, **kwargs)
 
 
 #
 # Private API
 #
 
-def _client(*args, **kwargs):
-    agent = kwargs.get('agent')
+def _client(kwargs):
```

Comment:
`args` was unused so I get dropping it to avoid losing args there but I'm not clear why `**kwargs` had to become `kwargs`.  Assuming it does have to change (I fully expect a good justification `:]`) I am a bit tempted to not use `kwargs` given the normal additional meaning of variables named that.  Or perhaps the `dict` aspect could be dropped entirely and just specify the args (`agent`, `pool`, `persistent`, `reactor`) explicitly here and above on `request()`?  I mean there are multiple lines of docstring for each parameter already so specifying it's name and default in the actual signature doesn't seem too crazy.  This `dict` mutation making the dict valid to be passed to another function outside this one doesn't smell great either.

The `HTTPClient.request()`-only args being `**kwargs` may still make sense but again, the docstring still reflects them so it's not like `request()` above is getting out of 'knowing' about the parameters.  If the docstring were changed to just refer and link to `HTTPClient.request()` then I think the `**kwargs` passthrough would make more sense.

Reply:
Basically, because this was supposed to be a tiny, very scoped PR :-).

But code-wise, it would make sense to re-think the whole flow of information. Issues #244, #143, even #87 are related to this.
The way things are being done now, everything gets passed around all the time, which is why the public interface accepts anything and never raises an Exception.

There is the `treq.client.HttpClient.request` method, which is the one that does actually does something, everything else just modifies some arguments to it. I think it makes sense, because that way adding some feature or fixing something is done in just that spot.

But that means that wherever there is a public API (e.g. `treq.get`), there is a `_client` call, which expects things like agent, reactor, ... and then that gets passed further to `HttpClient.request`.
So there were/are two options: change the signature of all those public methods to be complete, with the downside that further improving treq becomes burdensome (change a bunch of signatures as opposed to just one), or, what is being done here, which is consume the `_client` specific arguments and don't pass them further down the line.

In any case, as @glyph mentioned the right thing is to first raise a warning for some versions, so I guess this PR will have to stay open until then.

Reply:
> consume the _client specific arguments and don't pass them further down the line

I don't understand how changing this from `**kwargs` to `kwargs` keeps this a tiny, very scoped PR or why it is needed at all.  Did I miss the explanation of that in your reply?  It seems that instead it requires deleting the `**` all those public methods.

Also, having `def _client(reactor, agent=None, pool=None, persistent=None, **kwargs): kwargs['reactor'] = reactor` would give you the same `kwargs` as popping/getting from `kwargs` as this PR is now doing. Since it's an internal api I think not enforcing the keyword args only and not ignoring positional args is ok.

I see now how while `request()` above does have a complete docstring and doesn't reduce burden much by not having explicit parameters, that is not the case for each other individual api function.  Anyways, the signature suggested above doesn't require any changes to those api functions, I don't think, but it does make `_client()` use more normal mechanisms for accepting parameters I think and the PR is already changing how those parameters were handled so it doesn't seem to be any creep in scope/size, just an alternative change.

As to raising a warning, can't `**kwargs` to `HTTPClient.request()` and raise a warning if it's not empty?  Then after 'awhile' just delete the `**kwargs` and the warning code.  Then we don't have to keep maintaining this PR for a few releases.

Reply:
> I don't understand how changing this from **kwargs to kwargs keeps this a tiny, very scoped PR

Because all methods where `_client` is used, have a `kwargs` that may or may not have `agent`, `pool` and `persistent` defined.

So either: we rely on `_client` being an internal function alter those `kwargs` by `pop`-ing the keyword arguments that are not supposed to be passed further down the road, notice that that relies on the kwargs argument being the same dict than in each of the methods.
The alternative would be to clean the `kwargs` dict in each method before `request`, which is likely way too much work for this particular thing and that time would be better spent dealing with the issues I mentioned.

Your suggestion of an alternative `_client` signature is not enough, since it'd still mean `agent` and `pool` are passed to `treq.HttpClient.request`, triggering the warning/exception.

The "burden" I mention is because `treq.get`, `treq.put`, `treq.post`, `treq.HttpClient.get`, `treq.HttpClient.put`, ... all have pretty much the exact same interface, so any changes have to be done consistently in all those places.
The fact that everything ends in `treq.HttpClient.request` does mean less work and less errors.

> As to raising a warning, can't **kwargs to HTTPClient.request() and raise a warning if it's not empty? Then after 'awhile' just delete the **kwargs and the warning code. Then we don't have to keep maintaining this PR for a few releases.

That's a good call.

Reply:
Good golly gee, I think I finally get it.  Thanks.  Also, I should have had `reactor=None`...  But wow, this is gross.  :]  The part about calling `x(p).y(p)` and having `x()` mutate `p` to make it correct for `y()`.  Brainstorming here:

```python
def request(method, url, **kwargs):
    client, filtered_kwargs = _client(kwargs)
    return client.request(method, url, **filtered_kwargs)


def _client(reactor=None, agent=None, pool=None, persistent=None, **kwargs):
    if agent is None:
        reactor = default_reactor(reactor)
        pool = default_pool(reactor, pool, persistent)
        agent = Agent(reactor, pool=pool)

    return HTTPClient(agent), dict(kwargs, reactor=reactor)
```

Yes, this makes _more_ change to the existing wrapper APIs than just removing `**` but I think it results in more reasonable code.  That said, the goofy `kwargs` mutation is fairly contained here...  so I think I'd say this may be for someone else to step in and provide some guidance.

Here's a full PR for reference.  https://github.com/CamilionEU/treq/pull/1 (tox is running against all envs now, I had forgotten I only checked py3 before pushing)

Reply:
Maybe I only mentioned on IRC that it was a terrible hack :-). I figured we'd deprecate those bits soon enough that it wouldn't be necessary to keep that around too long, so I didn't think about it too much.

That's much better though :-) so it's fully merged.